File: sdks/dotnet/Condux.Sdk/ConduxClient.cs
using System.Text.Json;
using System.Text.Json.Serialization;

namespace Condux.Sdk;

/// <summary>
/// Reports errors to a Condux relay. Emits the Sentry "store" wire shape so the relay normalizes it exactly
/// like an official Sentry SDK — point it at a project DSN and it works. Delivery is resilient (429 / 5xx /
/// network failures are retried with backoff, honoring <c>Retry-After</c>) and <b>never throws</b>: a failed
/// send returns a <see cref="SendResult"/> rather than crashing the host app. One client is cheap to hold
/// for the app's lifetime. Inspired by common SDK transports, implemented fresh.
/// </summary>
public sealed class ConduxClient
{
    private static readonly JsonSerializerOptions Json = new()
    {
        PropertyNamingPolicy = JsonNamingPolicy.SnakeCaseLower,
        DefaultIgnoreCondition = JsonIgnoreCondition.WhenWritingNull,
    };

    private readonly ConduxOptions options;
    private readonly HttpClient http;
    private readonly Func<DateTimeOffset> clock;
    private readonly Func<TimeSpan, CancellationToken, Task> sleep;
    private readonly string storeUrl;
    private readonly string publicKey;

    /// <summary>Create a client for the given options. Throws on a malformed DSN (a config error that
    /// should fail loud at startup, not be swallowed per capture).</summary>
    public ConduxClient(ConduxOptions options)
    {
        this.options = options;
        var dsn = Dsn.Parse(options.Dsn);
        storeUrl = $"{dsn.Endpoint}/api/{dsn.ProjectId}/store/";
        publicKey = dsn.PublicKey;
        http = new HttpClient(options.Transport ?? new HttpClientHandler());
        clock = options.Clock ?? (() => DateTimeOffset.UtcNow);
        sleep = options.Sleep ?? Task.Delay;
    }

    /// <summary>
    /// Report an exception as an error-level event, with its stack trace. Pass <paramref name="handled"/>
    /// = false for an uncaught exception (a framework integration does this) so the relay marks it unhandled.
    /// </summary>
    public Task<SendResult> CaptureExceptionAsync(
        Exception error, bool handled = true, CancellationToken cancellationToken = default) =>
        DispatchAsync(Level.Error, message: null, error, handled, context: null, cancellationToken);

    /// <summary>
    /// Report an exception with detail belonging to this one event. <paramref name="context"/> carries the
    /// request it happened during and any request-scoped tags; they are passed here rather than set on
    /// <see cref="ConduxScope"/> because scope state outlives the call, so on a server handling requests
    /// concurrently request detail set ambiently attaches to whichever event is captured next.
    /// </summary>
    /// <remarks>An overload rather than an optional parameter, so existing calls that pass a
    /// CancellationToken positionally keep compiling and keep meaning the same thing.</remarks>
    public Task<SendResult> CaptureExceptionAsync(
        Exception error, bool handled, CaptureContext? context,
        CancellationToken cancellationToken = default) =>
        DispatchAsync(Level.Error, message: null, error, handled, context, cancellationToken);

    /// <summary>Report a bare message event at the given level (default <see cref="Level.Info"/>).</summary>
    public Task<SendResult> CaptureMessageAsync(
        string message, Level level = Level.Info, CancellationToken cancellationToken = default) =>
        DispatchAsync(level, message, error: null, handled: true, context: null, cancellationToken);

    /// <summary>Report a message with per-event detail. See the CaptureExceptionAsync overload.</summary>
    public Task<SendResult> CaptureMessageAsync(
        string message, Level level, CaptureContext? context,
        CancellationToken cancellationToken = default) =>
        DispatchAsync(level, message, error: null, handled: true, context, cancellationToken);

    // Reading the exception happens inside the guard below, not at the call site: a custom exception type
    // decides what its own Message and stack yield, so that read is one of the things that can fail.
    private async Task<SendResult> DispatchAsync(
        Level level, string? message, Exception? error, bool handled, CaptureContext? context,
        CancellationToken cancellationToken)
    {
        try
        {
            var exception = error is null ? null : StackTraceReader.ToException(error, handled);
            var scope = ConduxScope.Fields();
            var payload = new EventPayload
            {
                EventId = Guid.NewGuid().ToString("N"),
                Timestamp = clock().ToUnixTimeMilliseconds() / 1000.0, // epoch seconds, the store convention
                Platform = "csharp",
                Level = level.ToString().ToLowerInvariant(),
                Environment = options.Environment,
                Release = options.Release,
                Message = message,
                Exception = exception is null ? null : new ExceptionEnvelope { Values = [exception] },
                User = scope.User,
                Request = context?.Request,
                // Merged over the ambient tags rather than replacing them, so a per-event tag cannot
                // silently drop the deployment-wide ones.
                Tags = MergeTags(scope.Tags, context?.Tags),
                Contexts = scope.Contexts,
                Breadcrumbs = scope.Breadcrumbs,
            };

            var body = JsonSerializer.Serialize(payload, Json);
            return await EventTransport.SendAsync(
                http, storeUrl, publicKey, body, options.MaxRetries, sleep, cancellationToken);
        }
        catch (OperationCanceledException) when (cancellationToken.IsCancellationRequested)
        {
            throw; // the caller's own cancellation, not a reporting failure
        }
        catch (Exception failure)
        {
            // Reporting never throws: an error monitor that throws turns a handled error into an unhandled
            // one in exactly the code path where someone is already dealing with a failure. Anything the
            // transport does not already handle (an unserializable scope value, a disposed handler) comes
            // back as a failed result instead. Attempts is 0 — the event never left the process.
            return new SendResult(false, 0, null, failure.Message);
        }
    }

    private static IReadOnlyDictionary<string, string>? MergeTags(
        IReadOnlyDictionary<string, string>? ambient, IReadOnlyDictionary<string, string>? perEvent)
    {
        if (perEvent is null || perEvent.Count == 0)
        {
            return ambient;
        }

        var merged = ambient is null ? [] : new Dictionary<string, string>(ambient);
        foreach (var (key, value) in perEvent)
        {
            merged[key] = value;
        }
        return merged;
    }
}

/// <summary>
/// Per-event enrichment, passed at the capture call rather than set ambiently, for anything derived from
/// a single request. See the CaptureExceptionAsync overload for why that distinction matters.
/// </summary>
public sealed record CaptureContext
{
    /// <summary>The request the event happened during.</summary>
    public ConduxRequest? Request { get; init; }

    /// <summary>Tags for this event only, merged over the ambient ones.</summary>
    public IReadOnlyDictionary<string, string>? Tags { get; init; }
}

// A parsed DSN: scheme://<publicKey>@<host>/<projectId>.
internal readonly record struct Dsn(string Endpoint, string ProjectId, string PublicKey)
{
    // Throws on anything that is not a whole DSN. A key-less or project-less URL parses as a Uri quite
    // happily, and a client built from one sends every event to an address that can only be rejected —
    // so the check belongs here, where it fails once at startup rather than silently per capture.
    public static Dsn Parse(string dsn)
    {
        var projectId = Uri.TryCreate(dsn, UriKind.Absolute, out var uri) ? uri.AbsolutePath.TrimStart('/') : "";
        if (uri is null || uri.UserInfo.Length == 0 || projectId.Length == 0)
        {
            // One message for every shape of wrong DSN, naming what a right one looks like. The key in a
            // DSN is the public ingest key, so echoing the value back is not a secret leak.
            throw new FormatException($"Condux: DSN must be scheme://<key>@<host>/<projectId>, got '{dsn}'.");
        }

        return new Dsn($"{uri.Scheme}://{uri.Authority}", projectId, uri.UserInfo);
    }
}

File: sdks/jvm/src/main/java/ai/condux/ConduxClient.java
package ai.condux;

import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.UUID;
import java.util.function.DoubleConsumer;
import java.util.function.DoubleSupplier;

/**
 * Condux SDK for the JVM — report errors to a Condux relay from Java, Kotlin, or Scala.
 *
 * <p>Emits the Sentry "store" wire shape (event_id, timestamp, level, exception.values[]) so the relay
 * normalizes it exactly like an official Sentry SDK: swap the DSN and it works. Delivery is resilient
 * (429 / 5xx / network failures retry with backoff, honoring Retry-After) and never throws — a failed
 * send returns a {@link SendResult} you can inspect. The transport, sleep, and clock are injectable via
 * the builder so backoff is exercised with no real network or timers. Zero dependencies: HTTP is
 * {@link java.net.http.HttpClient} ({@link EventTransport}) and the JSON is written by hand ({@link Json}).
 *
 * <pre>{@code
 * ConduxClient condux = ConduxClient.builder("https://<key>@ingest.condux.ai/<projectId>")
 *     .environment("production")
 *     .release("1.4.2")
 *     .build();
 * try {
 *     doWork();
 * } catch (Exception e) {
 *     condux.captureException(e);
 *     throw e;
 * }
 * }</pre>
 */
public final class ConduxClient {
    private final EventTransport transport;
    private final String environment;
    private final String release;
    private final DoubleSupplier clock;

    private ConduxClient(Builder builder) {
        Dsn dsn = Dsn.parse(builder.dsn);
        Map<String, String> headers = Map.of("Content-Type", "application/json", "x-condux-auth", dsn.publicKey());
        Transport http = builder.transport != null ? builder.transport : EventTransport.httpClient();
        DoubleConsumer sleeper = builder.sleeper != null ? builder.sleeper : EventTransport::defaultSleep;
        this.transport = new EventTransport(dsn.storeUrl(), headers, builder.maxRetries, http, sleeper);
        this.environment = builder.environment;
        this.release = builder.release;
        this.clock = builder.clock != null ? builder.clock : () -> System.currentTimeMillis() / 1000.0;
    }

    /** Start building a client for the given project DSN. */
    public static Builder builder(String dsn) {
        return new Builder(dsn);
    }

    /** Report an exception as an error-level event, with its stack trace. Never throws on delivery failure. */
    public SendResult captureException(Throwable error) {
        return captureException(error, true);
    }

    /**
     * Report an exception, marking whether it was {@code handled}. A framework adapter that catches an
     * uncaught request exception reports {@code handled = false} (drives the unhandled badge).
     */
    public SendResult captureException(Throwable error, boolean handled) {
        Map<String, Object> fields = new LinkedHashMap<>();
        fields.put("level", Level.ERROR.wire());
        fields.put("exception", Map.of("values", List.of(EventPayload.exception(error, handled))));
        return dispatch(fields);
    }

    /** Report a bare message event at the given level. */
    public SendResult captureMessage(String message, Level level) {
        Map<String, Object> fields = new LinkedHashMap<>();
        fields.put("level", level.wire());
        fields.put("message", message);
        return dispatch(fields);
    }

    private SendResult dispatch(Map<String, Object> fields) {
        Map<String, Object> event = new LinkedHashMap<>();
        event.put("event_id", UUID.randomUUID().toString().replace("-", "")); // 32 lowercase hex
        event.put("timestamp", clock.getAsDouble());                          // epoch seconds
        event.put("platform", "java");
        event.putAll(ConduxScope.fields());
        event.putAll(fields);
        if (environment != null) {
            event.put("environment", environment);
        }
        if (release != null) {
            event.put("release", release);
        }
        return transport.send(Json.write(event));
    }

    /** Fluent builder; the DSN is required, everything else is optional (and testing hooks). */
    public static final class Builder {
        private final String dsn;
        private String environment;
        private String release;
        private int maxRetries = 3;
        private Transport transport;
        private DoubleConsumer sleeper;
        private DoubleSupplier clock;

        private Builder(String dsn) {
            this.dsn = dsn;
        }

        public Builder environment(String environment) {
            this.environment = environment;
            return this;
        }

        public Builder release(String release) {
            this.release = release;
            return this;
        }

        public Builder maxRetries(int maxRetries) {
            this.maxRetries = maxRetries;
            return this;
        }

        /** Override the HTTP transport (tests inject a scripted one). */
        public Builder transport(Transport transport) {
            this.transport = transport;
            return this;
        }

        /** Override the backoff sleep (tests record the delays instead of waiting). */
        public Builder sleeper(DoubleConsumer sleeper) {
            this.sleeper = sleeper;
            return this;
        }

        /** Override the event clock (tests pin the timestamp). */
        public Builder clock(DoubleSupplier clock) {
            this.clock = clock;
            return this;
        }

        public ConduxClient build() {
            return new ConduxClient(this);
        }
    }
}
